lib/fluent/plugin/in_tcp.rb

```diff
@@ -83,7 +85,32 @@ def start
               router.emit(tag, time, record)
             end
           end
-          conn.buffer.slice!(0, pos) if pos > 0
+          buf.slice!(0, pos) if pos > 0
+        end
+      else
+        server_create(:in_tcp_server_batch_emit, @port, bind: @bind, resolve_name: !!@source_hostname_key) do |data, conn|
+          conn.buffer << data
```

Comment:
Is it needed if using a local variable(`buf`)? #buffer is just a string. It is used only for buffering https://github.com/fluent/fluentd/blob/5a7d94941371dddcb7150a73ddbe142102a68964/lib/fluent/plugin_helper/server.rb#L448.

Reply:
`buf = conn.buffer` is for avoiding `buffer` accessor call in `while`.

Reply:
My question is `conn.buffer << data` is needed? `conn.buffer` will not be used in the following logic.

Reply:
Yes. `conn.buffer` may have previous data so need to concat before use.

Reply:
ah, I see 👍